Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Spawnable/InMemorySpawnableAssetContainer.cpp

```diff
@@ -213,63 +212,91 @@ namespace AzToolsFramework::Prefab::PrefabConversionUtils
         return m_spawnableAssets;
     }
 
-    void InMemorySpawnableAssetContainer::LoadReferencedAssets(AZStd::vector<AZ::Data::Asset<AZ::Data::AssetData>>& referencedAssets)
+    void InMemorySpawnableAssetContainer::LoadReferencedAssets(SpawnableAssetData& spawnable)
     {
-        // Start our loads on all assets by calling GetAsset from the AssetManager
-        for (AZ::Data::Asset<AZ::Data::AssetData>& asset : referencedAssets)
-        {
-            if (!asset.GetId().IsValid())
-            {
-                AZ_Error("Prefab", false, "Invalid asset found referenced in scene while entering game mode");
-                continue;
-            }
-
-            const AZ::Data::AssetLoadBehavior loadBehavior = asset.GetAutoLoadBehavior();
-
-            if (loadBehavior == AZ::Data::AssetLoadBehavior::NoLoad)
-            {
-                continue;
-            }
+        // Get the referenced assets directly from the product. This is done for two reasons:
+        //  1. Avoids calls to the Asset Manager for assets that are already loaded.
+        //  2. Gets the exact asset to load to avoid issues with assets that don't reload.
+        AZStd::vector<AZ::Data::Asset<AZ::Data::AssetData>*> blockingAssets;
 
-            AZ::Data::AssetId assetId = asset.GetId();
-            AZ::Data::AssetType assetType = asset.GetType();
-
-            asset = AZ::Data::AssetManager::Instance().GetAsset(assetId, assetType, loadBehavior);
-
-            if (!asset.GetId().IsValid())
+        for (AZ::Data::Asset<AZ::Data::AssetData>& asset : spawnable.m_assets)
+        {
+            AZ::SerializeContext* sc = nullptr;
+            AZ::ComponentApplicationBus::BroadcastResult(sc, &AZ::ComponentApplicationBus::Events::GetSerializeContext);
+            AZ_Assert(
+                sc, "Unable to locate Serialize Context while resolving asset references in the in-memory spawnable asset container.");
+
+            auto callback = [&blockingAssets](
+                                void* object, const AZ::SerializeContext::ClassData* classData,
+                                [[maybe_unused]]const AZ::SerializeContext::ClassElement* elementData) -> bool
             {
-                AZ_Error("Prefab", false, "Invalid asset found referenced in scene while entering game mode");
-                continue;
-            }
+                if (classData->m_typeId == AZ::GetAssetClassId())
+                {
+                    auto asset = reinterpret_cast<AZ::Data::Asset<AZ::Data::AssetData>*>(object);
+
+                    if (!asset->GetId().IsValid())
+                    {
+                        AZ_Error("Prefab", false, "Invalid asset found referenced in scene while entering game mode");
```

Comment:
+1

Reply:
If the id on the asset is invalid it typically means that the entire asset isn't valid so any retrieved data is empty or in a worse case misleading data. That said a bit lower there are two more errors that both were extended to include the asset hint, asset id and asset type.

Reply:
Yeah, since the asset was invalid it could be entirely deleted. I was more wondering if we could print the object trying to use the asset? For example, if a prefab, has a script component which is holding ref to a non-existing script asset, could we print the name of the prefab (or even the name of the entity in the prefab) which is storing the bad ref? 🤷‍♀️ maybe we don’t have that info at this point

Reply:
The owning Spawnable is an in-memory spawnable so it has a placeholder name. Printing that out would add more confusion than it would clarify I think. I extended the message to include the class name that owns the asset, so at least that will help a bit with narrowing down any issues hopefully.

Reply:
Ah, that makes sense. Thanks @AMZN-koppersr !